src/inference/nuts.jl

```diff
@@ -0,0 +1,191 @@
+using LinearAlgebra: dot
+
+struct Tree
+  val_left
+  momenta_left
+  val_right
+  momenta_right
+  val_sample
+  n :: Int
+  weight :: Float64
+  stop :: Bool
+  diverging :: Bool
+end
```

Comment:
For a structure like this (which I assume has recursive references to instances of itself) -- do you mind adding documentation to indicate this?

Alternatively, you might use an abstract type to directly indicate the recursion.